### ISSUE TITLE
OpenCL RNG

### ISSUE THREAD

**Issue by maedoc** [maedoc/sddekit#139]
We'll need to generate random numbers preferably not hand-coded; PyOpenCL uses [`RANLUX`](https://bitbucket.org/ivarun/ranluxcl/).

**Comment by catchchaos** [comment]
Consider combining a few primitive cyclic  generators as described [here](http://www.drdobbs.com/tools/fast-high-quality-parallel-random-number/229625477). This is results in a fast RNG that has independent sub generators and should be easy enough to implement in OpenCL.

**Comment by abheekg** [comment]
@catchchaos We are using AMD's clRNG library.

**Comment by maedoc** [comment]
I would prefer to choose one or two with the highest periods, which is an important factor in quality:

- RANLUX has a period of 10^171 (as described in the original paper [PDF](http://arxiv.org/pdf/hep-lat/9309020v1.pdf)),
- The best one in the Dobbs article has a period of 10^14
- MRG31k3p has a period of 2^185, or 10^55 ([source](http://simul.iro.umontreal.ca/ssj/doc/html/umontreal/iro/lecuyer/rng/MRG31k3p.html))
- MRG32k3a is 10^57 ([source](https://software.intel.com/en-us/node/590402))
- LFSR113 is 10^34 ([source](http://simul.iro.umontreal.ca/ssj/doc/html/umontreal/iro/lecuyer/rng/LFSR113.html))

Imagine a parameter sweep of 5 parameters, each having 100 values, and each simulation requires 10000000 time steps. If there are 400 state variables, we will need around 10^20 random numbers.

Because OpenCL targets hardware with single precision, there's no reason not to choose RANLUX, though again having a second one to compare with could be good.

**Comment by catchchaos** [comment]
Yup, RANLUX looks like our best bet currently and is certainly one of the popular options. If you're still looking for alternatives, I did come across the MIXMAX RNG described in [this paper](http://arxiv.org/pdf/1403.5355v2.pdf). I'll look at it more thoroughly if you find it promising.

**Comment by catchchaos** [comment]
An OpenCl implementation of RANLUX PRNG in OpenCl can be found [here](https://bitbucket.org/ivarun/ranluxcl/overview).

**Comment by maedoc** [comment]
yup that's definitely the one linked to in the description of this issue.

**Comment by catchchaos** [comment]
Oops sorry I didn't see that :p
I just happened to come across it while looking for a PRNG and thought I'd mention it for further reference.
